### PR TITLE
add AlertsAgent

### PR DESCRIPTION
**Pull request by webispy**
add AlertsAgent to support ALARM/TIMER.

Signed-off-by: Inho Oh <inho.oh@sk.com>

### CODE REVIEW

File: src/alerts_manager.hh [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    void reset();
+
+    void done(AlertItem* item);
+
+    void activate(AlertItem* item);
+    void deactivate(AlertItem* item);
+
+    void dump();
+
+    std::map<std::string, AlertItem*> token_map;
```
Comment (kimhyungrok) [inline src/alerts_manager.hh:114]: Please consider to change private access specifier.

File: include/alerts_agent.hh [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+#include <clientkit/capability.hh>
+
+#include "alerts_audio_player.hh"
+#include "base_audio_player_listener.hh"
+
+#include <glib.h>
+#include <json/json.h>
+
+using namespace NuguClientKit;
```
[the remark this answers is not in the document]
Comment (webispy) [reply include/alerts_agent.hh:28]: Thank you! NuguClientKit is necessary. I will change the `NuguClientKit::FocusState` to `FocusState`

File: include/alerts_agent.hh [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    /* IAlertsManagerListener */
+    void onTimeout(const std::string& token) override;
+    void onAssetRequireTimeout(const std::string& token) override;
+    void onDurationTimeout(const std::string& token) override;
+
+    NuguCapability::AlertsAudioPlayer* audioplayer;
+    std::string playstackctl_ps_id;
+    NuguClientKit::FocusState focus_state;
+    NuguDirective* directive_for_sync;
+    NuguClientKit::IPlaySyncManager* playsync_manager;
```
[the remark this answers is not in the document]
Comment (webispy) [reply include/alerts_agent.hh:139]: Thank you! I will remove the `playsync_manager`

File: src/alerts_manager.hh (hunk 1 above)
Comment (webispy) [reply src/alerts_manager.hh:114]: Thank you! I will move `token_map` to private

File: src/alerts_agent.cc [hunk 4]
```diff
@@ -0,0 +1,841 @@
...
+    delete manager;
+}
+
+void AlertsAgent::initialize()
+{
+    if (initialized) {
+        nugu_warn("It's already initialized.");
+        return;
+    }
+
```
Comment (kimhyungrok) [inline src/alerts_agent.cc:72]: There needs this code [Capability::initialize();] in this location, if you use parent Capability.

File: src/alerts_agent.cc [hunk 5]
```diff
@@ -0,0 +1,841 @@
...
+    addReferrerEvents("SetSnoozeFailed", "SetSnooze");
+    addReferrerEvents("SetAlertAssetSucceeded", "DeliveryAlertAsset");
+    addReferrerEvents("SetAlertAssetFailed", "DeliveryAlertAsset");
+
+    audioplayer = new AlertsAudioPlayer();
+    audioplayer->setNuguCoreContainer(core_container);
+    audioplayer->initialize();
+    audioplayer->addListener(this);
+
+    playsync_manager = capa_helper->getPlaySyncManager();
```
Comment (kimhyungrok) [inline src/alerts_agent.cc:87]: This is done by parent Capabilty already. So, it's not necessary this. (It's all equivalent to focus_manager, session_manager, ... etc.)

File: src/alerts_agent.cc (hunk 4 above)
Comment (webispy) [reply src/alerts_agent.cc:72]: Thank you! I will add the `Capability::initialize();` to `AlertsAgent::initialize()` function

File: src/alerts_agent.cc (hunk 5 above)
Comment (webispy) [reply src/alerts_agent.cc:87]: Thank you! I will remove the line 87.

File: src/alerts_agent.cc [hunk 6]
```diff
@@ -0,0 +1,829 @@
...
+
+    nugu_dbg("remove %d alarms", tokens.size());
+
+    for (int i = 0; i < (int)tokens.size(); i++) {
+        std::string token = tokens[i].asString();
+
+        nugu_info("remove %d/%d: %s", i + 1, tokens.size(), token.c_str());
+
+        if (alerts_listener)
+            alerts_listener->onAlertDelete(token);
```
Comment (JeanTracker) [inline src/alerts_agent.cc:405]: It's weired that the event is notified before remove alerts.

**Review by kimhyungrok — APPROVED** [review]

File: src/alerts_agent.cc [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+                use_file = true;
+        } else if (current->rsrc_type == "TTS") {
+            if (audioplayer->playTTS() == false)
+                use_file = true;
+        } else if (current->rsrc_type == "INTERNAL") {
+            use_file = true;
+        } else {
+            nugu_error("unknown resource type: %s", current->rsrc_type.c_str());
+            use_file = true;
+        }
```
Comment (JeanTracker) [inline src/alerts_agent.cc:774]: `use_file` is `true` in most cases. How about changing to conditional code where it is `false`?

File: src/alerts_agent.cc (hunk 6 above)
Comment (webispy) [reply src/alerts_agent.cc:405]: Thank you! I will move to the next line of `removeAlert()`

File: src/alerts_agent.cc [hunk 8]
```diff
@@ -0,0 +1,841 @@
...
+using namespace NuguCapability;
+
+static const char* CAPABILITY_NAME = "Alerts";
+static const char* CAPABILITY_VERSION = "1.1";
+
+AlertsAgent::AlertsAgent()
+    : Capability(CAPABILITY_NAME, CAPABILITY_VERSION)
+    , playstackctl_ps_id("")
+    , manager(new AlertsManager())
+    , current(nullptr)
```
[the remark this answers is not in the document]
Comment (webispy) [reply src/alerts_agent.cc:51]: Thank you. I will add `audioplyaer(nullptr)` to constructor.

File: src/alerts_agent.cc (hunk 7 above)
Comment (webispy) [reply src/alerts_agent.cc:774]: Ok. good suggestion. I will change the default value of `use_file` to `true`

File: include/alerts_agent.hh [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+    IAlertsListener* alerts_listener = nullptr;
+
+    bool holdAlarmByText;
```
[the remark this answers is not in the document]
Comment (webispy) [reply include/alerts_agent.hh:142]: OK. I will change to `hold_alarm_by_test`

**Pull request merged**